asapdiscovery-modeling/asapdiscovery/modeling/scripts/blast_search.py

```diff
@@ -0,0 +1,129 @@
+
+    parser.add_argument(
+        "-if",
+        "--in_fasta",
+        required=True,
+        help="Protein FASTA sequence to search in BLAST.",
+    )
+    parser.add_argument(
+        "-of",
+        "--out_fasta",
+        help="FASTA file for BLAST results.",
+    )
+    parser.add_argument(
+        "-o", "--out_fn", required=True, help="Output CSV file name."
+    )
+    parser.add_argument(
+        "--cache", help="Optional cache file to save raw XML results."
+    )
+    parser.add_argument(
+        "--overwrite",
+        action="store_true",
+        help="Overwrite cache file if it already exists.",
+    )
+
+    parser.add_argument(
+        "-n",
+        "--n_hits",
+        type=int,
+        default=500,
+        help="Number of hits to return.",
+    )
+    parser.add_argument(
+        "-e",
+        "--e_val_thresh",
+        type=float,
+        default=1e-20,
+        help="Threshold for BLAST E value.",
+    )
+    parser.add_argument(
+        "-f",
+        "--filter",
+        nargs="+",
+        default=[],
+        help="Regex(es) to search in hit title.",
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    args = get_args()
+
+    # Check for cache and load if present
+    if args.cache and os.path.isfile(args.cache) and (not args.overwrite):
+        print("Loading from cached download", flush=True)
+        result_handle = open(args.cache)
+    else:
+        print("Searching BLASTP", flush=True)
+        # Load reference sequence
+        fasta_seq = open(args.in_fasta).read().strip()
+
+        # Run BLASTP and get results
+        result_handle = NCBIWWW.qblast(
+            program="blastp",
+            database="refseq_protein",
+            sequence=fasta_seq,
+            hitlist_size=args.n_hits,
+            expect=args.e_val_thresh,
+        )
+
+        # Save BLAST results
+        if args.cache:
+            print("Saving cache", flush=True)
+            with open(args.cache, "w") as fp:
+                fp.write(result_handle.read())
+            result_handle.seek(0)
+
+    # Parse BLAST results
+    result_seqs = {}
+    total_results = 0
+    for record in NCBIXML.parse(result_handle):
+        if record.alignments:
+            for align in record.alignments:
+                total_results += 1
+                keep_aln = False
+                if len(args.filter) == 0:
+                    keep_aln = True
+                for f in args.filter:
+                    if re.search(f, align.title):
+                        keep_aln = True
+                        break
+
+                if keep_aln:
+                    # Save sequence identity, title, and gapless sequence
+                    #  substring that aligns
+                    sequence_to_model = align.hsps[0].sbjct.replace("-", "")
+                    result_seqs[align.title] = sequence_to_model
+    print(f"Kept {len(result_seqs)} out of {total_results} hits", flush=True)
```

Comment:
just from a readability/debuggability perspective it might be better to restructure/break this up into something like this:

```
# Parse BLAST results
result_seqs = {}
total_results = 0
keep_aln = False # will switch to True if any filter flags are found in records.
for record in NCBIXML.parse(result_handle):
    if not record.alignments:
        continue # skip records if no alignments found

    for align in record.alignments:
        total_results += 1
        if len(args.filter) == 0:
            keep_aln = True

        if any(re.search(f, align.title) for f in args.filter):
            keep_aln = True

        if keep_aln:
            # Save sequence identity, title, and gapless sequence
            # substring that aligns
            sequence_to_model = align.hsps[0].sbjct.replace("-", "")
            result_seqs[align.title] = sequence_to_model

print(f"Kept {len(result_seqs)} out of {total_results} hits", flush=True)
```
Note: I didn't test this code - the `any()` call might not work as intended in its current form.

Reply:
done in [072157b](https://github.com/choderalab/asapdiscovery/pull/141/commits/072157ba51f5daf3e4332fba0d8a6b3fefe01077)